dotnet/src/InternalUtilities/src/Diagnostics/ModelDiagnostics.cs

```diff
@@ -0,0 +1,280 @@
+
+    /// <summary>
+    /// Set the response id for a given activity.
+    /// </summary>
+    /// <param name="activity">The activity to set the response id</param>
+    /// <param name="responseId">The response id</param>
+    /// <returns>The activity with the response id set for chaining</returns>
+    public static Activity SetResponseId(this Activity activity, string responseId) => activity.SetTag(ModelDiagnosticsTags.ResponseId, responseId);
+
+    /// <summary>
+    /// Set the prompt token usage for a given activity.
+    /// </summary>
+    /// <param name="activity">The activity to set the prompt token usage</param>
+    /// <param name="promptTokens">The number of prompt tokens used</param>
+    /// <returns>The activity with the prompt token usage set for chaining</returns>
+    public static Activity SetPromptTokenUsage(this Activity activity, int promptTokens) => activity.SetTag(ModelDiagnosticsTags.PromptToken, promptTokens);
+
+    /// <summary>
+    /// Set the completion token usage for a given activity.
+    /// </summary>
+    /// <param name="activity">The activity to set the completion token usage</param>
+    /// <param name="completionTokens">The number of completion tokens used</param>
+    /// <returns>The activity with the completion token usage set for chaining</returns>
+    public static Activity SetCompletionTokenUsage(this Activity activity, int completionTokens) => activity.SetTag(ModelDiagnosticsTags.CompletionToken, completionTokens);
+
+    # region Private
+    /// <summary>
+    /// Check if model diagnostics is enabled
+    /// Model diagnostics is enabled if either EnableModelDiagnostics or EnableSensitiveEvents is set to true and there are listeners.
+    /// </summary>
+    private static bool IsModelDiagnosticsEnabled()
+    {
+        return (s_enableDiagnostics || s_enableSensitiveEvents) && s_activitySource.HasListeners();
+    }
+
+    private static void AddOptionalTags(Activity? activity, PromptExecutionSettings? executionSettings)
+    {
+        if (activity is null || executionSettings?.ExtensionData is null)
+        {
+            return;
+        }
+
+        void TryAddTag(string key, string tag)
+        {
+            if (executionSettings.ExtensionData.TryGetValue(key, out var value))
+            {
+                activity.SetTag(tag, value);
+            }
+        }
+
+        TryAddTag("max_tokens", ModelDiagnosticsTags.MaxToken);
+        TryAddTag("temperature", ModelDiagnosticsTags.Temperature);
+        TryAddTag("top_p", ModelDiagnosticsTags.TopP);
+    }
+
+    /// <summary>
+    /// Convert chat history to a string aligned with the OpenAI format
+    /// </summary>
+    private static string ToOpenAIFormat(IEnumerable<ChatMessageContent> chatHistory)
+    {
+        return $"[{string.Join(
+            ", \n",
+            chatHistory.Select(m => $"{{\"role:\" {m.Role}, \"content:\" {m.Content}}}"))}]";
```

Comment:
It'd be nice to avoid all these intermediate allocations and instead format into a StringBuilder.

Reply:
Updated with the optimization. However, we cannot avoid the intermediate string objects completely as I realized I had to make them Json compatible (the content may contain line breaks).